src/components/cl/basic/cl_basic_lib.c

```diff
@@ -52,42 +53,42 @@ static inline ucc_status_t check_tl_lib_attr(const ucc_base_lib_t *lib,
 ucc_status_t ucc_cl_basic_get_lib_attr(const ucc_base_lib_t *lib,
                                        ucc_base_lib_attr_t  *base_attr)
 {
-    ucc_cl_lib_attr_t *attr   = ucc_derived_of(base_attr, ucc_cl_lib_attr_t);
-    ucc_cl_lib_t *     cl_lib = ucc_derived_of(lib, ucc_cl_lib_t);
-    ucc_config_names_array_t *tls = &cl_lib->tls;
-    ucc_tl_iface_t *          tl_iface;
-    int                       i;
-    ucc_status_t              status;
+    ucc_cl_lib_attr_t  *attr     = ucc_derived_of(base_attr, ucc_cl_lib_attr_t);
+    ucc_cl_basic_lib_t *cl_lib   = ucc_derived_of(lib, ucc_cl_basic_lib_t);
+    ucc_config_allow_list_t *tls = &cl_lib->super.tls;
+    ucc_tl_iface_t          *tl_iface;
+    int                      i;
+    ucc_status_t             status;
 
-    attr->tls                    = &cl_lib->tls;
+    attr->tls                = &cl_lib->super.tls.array;
+    cl_lib->tls_forced.count = 0;
+    cl_lib->tls_forced.names = NULL;
+    if (cl_lib->super.tls.mode == UCC_CONFIG_ALLOW_LIST_ALLOW) {
+        status = ucc_config_names_array_dup(&cl_lib->tls_forced,
+                                            &cl_lib->super.tls.array);
+        if (UCC_OK != status) {
+            return status;
+        }
+    }
+    attr->tls_forced             = &cl_lib->tls_forced;
     attr->super.attr.thread_mode = UCC_THREAD_MULTIPLE;
     attr->super.attr.coll_types  = 0;
     attr->super.flags            = 0;
-    if (tls->count == 1 && !strcmp(tls->names[0], "all")) {
-        /* Check all available components, since CL_BASIC_TLS == "all" */
-        for (i = 0; i < ucc_global_config.tl_framework.n_components; i++) {
-            tl_iface = ucc_derived_of(
-                ucc_global_config.tl_framework.components[i], ucc_tl_iface_t);
-            ucc_assert(tl_iface);
-            if (UCC_OK != (status = check_tl_lib_attr(lib, tl_iface, attr))) {
-                return status;
-            }
+
+    ucc_assert(tls->array.count >= 1);
+    for (i = 0; i < tls->array.count; i++) {
```

Comment:
this can be added to super class constructor if we move tls_forced to base class

Reply:
this is different for different CLs. CL/HIER builds tls_forced list based on the corresponding TL lists of SBGP subgroups. hence, can't be moved to common superclass